core/src/worker/workflow/machines/patch_state_machine.rs

```diff
@@ -71,53 +89,91 @@ pub(super) enum PatchCommand {}
 /// are guaranteed to return the same value.
 /// `replaying_when_invoked`: If the workflow is replaying when this invocation occurs, this needs
 /// to be set to true.
-pub(super) fn has_change(
+pub(super) fn has_change<'a>(
     patch_id: String,
     replaying_when_invoked: bool,
     deprecated: bool,
-) -> NewMachineWithCommand {
-    let (machine, command) =
-        PatchMachine::new_scheduled(SharedState { patch_id }, replaying_when_invoked, deprecated);
-    NewMachineWithCommand {
-        command,
-        machine: machine.into(),
-    }
-}
+    seen_in_peekahead: bool,
+    existing_patch_ids: impl Iterator<Item = &'a str>,
+    internal_flags: InternalFlagsRef,
+) -> Result<(NewMachineWithCommand, Vec<MachineResponse>), WFMachinesError> {
+    let shared_state = SharedState { patch_id };
+    let initial_state = if replaying_when_invoked {
+        Replaying {}.into()
+    } else {
+        Executing {}.into()
+    };
+    let command = Command {
+        command_type: CommandType::RecordMarker as i32,
+        attributes: Some(
+            RecordMarkerCommandAttributes {
+                marker_name: PATCH_MARKER_NAME.to_string(),
+                details: build_has_change_marker_details(&shared_state.patch_id, deprecated),
+                header: None,
+                failure: None,
+            }
+            .into(),
+        ),
+    };
+    let mut machine = PatchMachine {
+        state: initial_state,
+        shared_state,
+    };
 
-impl PatchMachine {
-    fn new_scheduled(
-        state: SharedState,
-        replaying_when_invoked: bool,
-        deprecated: bool,
-    ) -> (Self, Command) {
-        let initial_state = if replaying_when_invoked {
-            Replaying {}.into()
+    OnEventWrapper::on_event_mut(&mut machine, PatchMachineEvents::Schedule)
+        .expect("Patch machine scheduling doesn't fail");
+
+    // Possibly produce an upsert SA command for this patch.
+    let maybe_upsert_cmd = if replaying_when_invoked && !seen_in_peekahead
+        || !internal_flags.borrow_mut().try_use(
+            CoreInternalFlags::UpsertSearchAttributeOnPatch,
+            !replaying_when_invoked,
+        ) {
+        // If we're replaying but this patch isn't in the peekahead, then we wouldn't have
+        // upserted either, and thus should not create the machine
+        vec![]
+    } else {
+        let mut all_ids = HashSet::<_, RandomState>::from_iter(existing_patch_ids);
```

Comment:
Is this set ordered by insertion? I think this should be a vec. Even if Elastic/visibility doesn't care, we should have user-predictable order of SA values.

Reply:
None of that really matters since this isn't how they're going to show up on the lang side when reading. That's up to the lang implementation.

Reply:
I think it matters because we should send the command in the same order every time. If we are asking people to be deterministic in their upsert we should too (and not just that it happens to be deterministic to iterate a set in Rust, but predictably deterministic).

Reply:
Changed